include/rmm/device_scalar.hpp

```diff
@@ -92,6 +110,30 @@ class device_scalar {
 
  private:
   rmm::device_buffer buff{sizeof(T)};
+
+  template<bool synchronize>
+  inline T _value(cudaStream_t stream) const {
+    T host_value{};
+    _memcpy<synchronize>(&host_value, buff.data(), sizeof(T), stream);
```

Comment:
I just realized that this may not be asynchronous even when `synchronize` is `false`. `cudaMemcpyAsync` is not asynchronous if the host memory is pageable.  https://docs.nvidia.com/cuda/cuda-runtime-api/api-sync-behavior.html#api-sync-behavior__memcpy-async "For transfers from device memory to pageable host memory, the function will return only once the copy has completed."

Is stack memory pageable?

Reply:
If this is *never* asynchronous, then there is no reason to make an async version of the `value()` method...

Reply:
I believe all memory allocated by the process on the CPU address space is virtual memory, and map down to the page table. I imagine the `cudaMemcpyAsync` has to do something similar as `cudaHostRegister` and copy the pageable memory into a pinned memory region before copying it to the target device.

Reply:
That's what I thought too.  So Chris can probably scrap all this complexity and just have only synchronous versions. Should probably still take an optional stream though to avoid synchronizing the default stream...

Reply:
Using `cudaMemcpyAsync` would still be beneficial if the source and target are device buffers though, right? The caveat being it inherits this pageable-memory synchronous limitation from `cudaMemcpyAsync`.

Reply:
Yes, but in this code, the source or target are always a scalar host variable.

Reply:
Woops, I just saw that this is explicitly only copying from host memory. I agree with Mark.

Reply:
It sounds like we can do without `cudaStreamSynchronize` then.